client/ipa-join.c

```diff
@@ -612,8 +617,246 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
     return rval;
 }
 
+#else
+
+size_t
+jsonrpc_handle_response(char *ptr, size_t size, size_t nmemb, void *userdata) {
+    size_t realsize = size * nmemb;
+    struct curl_buffer *p = (struct curl_buffer *) userdata;
+
+    p->payload = (char *) realloc(p->payload, p->size + realsize + 1);
+
+    if (p->payload == NULL) {
+        fprintf(stderr, _("Expanding buffer in jsonrpc_handle_response failed"));
+        free(p->payload);
+        return 1;
+    }
+
+    memcpy(&(p->payload[p->size]), ptr, realsize);
+
+    p->size += realsize;
+    p->payload[p->size] = 0;
+
+    return realsize;
+}
+
 static int
-unenroll_host(const char *server, const char *hostname, const char *ktname, int quiet)
+join_krb5_jsonrpc(const char *ipaserver, char *hostname, char **hostdn, const char **princ, bool force, bool quiet) {
```

Comment:
This function is long and uses a lot of variables. Both make it harder to review the function and track down memory leaks. Could you please split the function into smaller helper functions, e.g. init curl, build the request, parse response, ...?

Reply:
Done. I hope... :wink: